File: apps/backend/src/portfolio/AssetTradingRecord.ts
import {
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { TradingType, UUID } from '@mid/shared';
import { PortfolioAsset } from './PortfolioAsset';
import { Account } from '../account/Account';

@Entity()
export class AssetTradingRecord {
  @PrimaryGeneratedColumn('uuid')
  id: UUID;

  @ManyToOne(() => Account, {
    lazy: true,
    nullable: true,
  })
  @JoinColumn({ name: 'account_id' })
  account: Account;

  @ManyToOne(() => PortfolioAsset, (portfolioAsset) => portfolioAsset.records, {
    nullable: false,
    eager: true,
  })
  @JoinColumn({ name: 'portfolio_asset_id' })
  portfolioAsset: PortfolioAsset;

  @Column({ type: 'varchar', length: 50 })
  ticker: string;

  @Column({ type: 'varchar', length: 50 })
  tradingType: TradingType;

  @Column({ type: 'bigint' })
  quantity: number;

  @Column({ type: 'bigint' })
  price: number;

  @Column({ type: 'date' })
  tradingDate: Date;
}

File: backend/src/main/kotlin/com/kyoulho/mid/portfolio/AssetTradingRecord.kt
package com.kyoulho.mid.portfolio

import com.kyoulho.mid.account.entity.Account
import jakarta.persistence.*
import java.util.*

@Entity
data class AssetTradingRecord(
        @Id
        @GeneratedValue(strategy = GenerationType.UUID)
        val id: String? = null,

        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "account_id")
        val account: Account,

        @ManyToOne(fetch = FetchType.EAGER, optional = false)
        @JoinColumn(name = "portfolio_asset_id", nullable = false)
        val portfolioAsset: PortfolioAsset,

        @Column(length = 50, nullable = false)
        val ticker: String,

        @Column(length = 50, nullable = false)
        @Enumerated(EnumType.STRING)
        val tradingType: TradingType,

        @Column(nullable = false)
        val quantity: Long,

        @Column(nullable = false)
        val price: Long,

        @Column(nullable = false)
        val tradingDate: Date
)
